application-task-default/src/main/java/com/xwiki/task/internal/notifications/taskchanged/TaskChangedEventListener.java

```diff
@@ -88,15 +93,40 @@ public void onEvent(Event event, Object source, Object data)
         WatchedLocationReference docRef =
             watchedEntityFactory.createWatchedLocationReference(taskChangedEvent.getDocument().getDocumentReference());
         // In order to receive notifications, watch the task page for the newly assigned user.
-        watchTask(docRef, (String) taskChangedEvent.getCurrentValue());
+        watchTask(docRef, (String) taskChangedEvent.getCurrentValue(), (XWikiContext) data);
         // In order to stop receiving notifications, unwatch the task page for the unassigned user.
-        unwatchTask(docRef, (String) taskChangedEvent.getPreviousValue());
+        unwatchTask(docRef, (String) taskChangedEvent.getPreviousValue(), (XWikiContext) data);
+    }
+
+    private boolean hasTaskNotificationPreferenceEnabled(DocumentReference user, XWikiContext context)
```

Comment:
Isn't this what you need here? https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-default/src/main/java/org/xwiki/notifications/notifiers/internal/UserEventManager.java#L112